src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java

```diff
@@ -1190,4 +1197,34 @@ public InputStream getFileContent(BitbucketSCMFile file) throws IOException, Int
         }
         return content;
     }
+
+    @Restricted(ProtectedExternally.class)
+    protected CloseableHttpResponse executeMethod(CloseableHttpClient client, HttpRequestBase httpMethod) throws IOException, InterruptedException {
+        CloseableHttpResponse response = client.execute(httpMethod, context);
+        Instant start = Instant.now();
+        Instant forcedEnd = start.plus(API_RATE_LIMIT_MAX_SLEEP);
+        Duration sleepDuration = API_RATE_LIMIT_INITIAL_SLEEP;
+        while (response.getStatusLine().getStatusCode() == API_RATE_LIMIT_STATUS_CODE
+                && Instant.now().plus(sleepDuration).isBefore(forcedEnd)) {
+            try {
+                response.close();
+            } finally {
+                httpMethod.releaseConnection();
+            }
+            /*
+             * TODO: If The Bitbucket Server API ever starts sending rate limit expiration time, we should
+             * change this to a more precise sleep.
+             * TODO: It would be better to log this to a context-appropriate TaskListener, e.g. an org/repo scan log.
+             */
+            LOGGER.log(Level.FINE, "Bitbucket server API rate limit reached, sleeping for {0} before retrying",
```

Comment:
A notable side-effect of not being able to log to a `TaskListener` here is that in things like org scan logs, there will be no indication that rate limiting is occurring, the log will just appear to hang. I do not see any way to improve that though without significant API changes, and the Bitbucket Cloud API client rate limit retrying has the same issue.